channel/funder.go

```diff
@@ -0,0 +1,231 @@
+// limitations under the License.
+
+package channel
+
+import (
+	"context"
+	"errors"
+	"github.com/stellar/go/keypair"
+	"github.com/stellar/go/xdr"
+	"log"
+	pchannel "perun.network/go-perun/channel"
+	"perun.network/perun-stellar-backend/channel/env"
+	"perun.network/perun-stellar-backend/wallet"
+	"perun.network/perun-stellar-backend/wire"
+	"time"
+)
+
+const MaxIterationsUntilAbort = 20
+const DefaultPollingInterval = time.Duration(6) * time.Second
+
+type Funder struct {
+	stellarClient   *env.StellarClient
+	acc             *wallet.Account
+	kpFull          *keypair.Full
+	perunID         xdr.ScAddress
+	assetID         xdr.ScAddress
+	maxIters        int
+	pollingInterval time.Duration
+}
+
+func NewFunder(acc *wallet.Account, kp *keypair.Full, stellarClient *env.StellarClient, perunID xdr.ScAddress, assetID xdr.ScAddress) *Funder {
+	return &Funder{
+		stellarClient:   stellarClient,
+		acc:             acc,
+		kpFull:          kp,
+		perunID:         perunID,
+		assetID:         assetID,
+		maxIters:        MaxIterationsUntilAbort,
+		pollingInterval: DefaultPollingInterval,
+	}
+}
+
+func (f *Funder) GetPerunID() xdr.ScAddress {
+	return f.perunID
+}
+
+func (f *Funder) GetAssetID() xdr.ScAddress {
+	return f.assetID
+}
+
+func (f *Funder) Fund(ctx context.Context, req pchannel.FundingReq) error {
+	log.Println("Fund called")
+	switch req.Idx {
+	case 0:
+		return f.fundPartyA(ctx, req)
+	case 1:
+		return f.fundPartyB(ctx, req)
+	default:
+		return errors.New("invalid index")
+	}
+}
+
+func (f *Funder) fundPartyA(ctx context.Context, req pchannel.FundingReq) error {
+
+	err := f.OpenChannel(ctx, req.Params, req.State)
+	if err != nil {
+
+		return errors.New("error while opening channel in party A")
+	}
+	err = f.FundChannel(ctx, req.Params, req.State, false)
+	if err != nil {
+		return err
+	}
+
+polling:
+	for i := 0; i < f.maxIters; i++ {
+		select {
+		case <-ctx.Done():
+			return f.AbortChannel(ctx, req.Params, req.State)
+		case <-time.After(f.pollingInterval):
+			chanState, err := f.GetChannelState(ctx, req.Params, req.State)
+			if err != nil {
+				continue polling
+			}
+			if chanState.Control.FundedA && chanState.Control.FundedB {
+				return nil
+			}
+
+		}
+	}
+	return f.AbortChannel(ctx, req.Params, req.State)
+}
+
+func (f *Funder) fundPartyB(ctx context.Context, req pchannel.FundingReq) error {
+
+polling:
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(f.pollingInterval):
+			log.Println("Party B: Polling for opened channel...")
+			chanState, err := f.GetChannelState(ctx, req.Params, req.State)
+			if err != nil {
+				log.Println("Party B: Error while polling for opened channel:", err)
+				continue polling
+			}
+			log.Println("Party B: Found opened channel!")
+			if chanState.Control.FundedA && chanState.Control.FundedB {
+				return nil
+			}
+			return f.FundChannel(ctx, req.Params, req.State, true)
```

Comment:
1. We should only call Fund, if this party's balance is not 0.
2. We shoud poll the channel state after funding, to verify that at some point both parties have funded. If not, we should abort to reclaim our funds.

With the current implementation, there are some problems. For example, if party A opens the channel but does not fund it, party B will fund it and never reclaim its funds using Abort.
Essentially, the funding should be veriy similar for both parties.

Rought protocol sketch:
1. Open the channel (Only Party A)
2. Poll until open channel is detected via GetChannelState
3. If Control.Funded[ME] is set, return
4. Call Fund
5. Poll GetChannelState until FundedA AND FundedB is true (e.g. 20 times), if reached, return
6. if polling ends without finding a funded channel, call abort.

There is no need to duplicate code and put 2 functions I think. wrap the OpenChannel in an `if req.Idx == 0`. The ckb backend is not a good reference here, as it has an inherently different funding protocol because of UTXO.